pipeline/pipelines/code_generation_pipeline.py

```diff
@@ -56,6 +60,13 @@ def _validate_codegen_kwargs(extra_args, **kwargs):
     _validate_toolkit_path(kwargs['toolkit_path'])
 
 
+def _validate_publish_kwargs(**kwargs):
+    if kwargs['publish_env'] not in _PYTHON_PUB_ENVS:
+        raise ValueError("Invalid devpi publishing environment " +
```

Comment:
- use single quotes
- prefer string format to concatentation; then you don't have to manually stringify.

E.g.,
'Invalid environment "{}"; choose from one of {}'.format(kwargs['publish_env'], _PYTHON_PUB_ENVS)